### PR TITLE
Make capsys more like stdio streams in python3. Resolves #1407.

### LINKED ISSUE #1407: capsys + sys.stdout.buffer => AttributeError
Applicable versions where sys.stdout.buffer should exist and be a bytestream:
- python3.x

``` python
import sys


def test(capsys):
    sys.stdout.buffer.write(b'hi\n')
    assert capsys.readouterr()[0] == 'hi\n'
```

```
$ py.test test.py
============================= test session starts ==============================
platform linux -- Python 3.4.3, pytest-2.8.7, py-1.4.31, pluggy-0.3.1
rootdir: /home/asottile/workspace/cheetah_lint, inifile: 
collected 1 items 

test.py F

=================================== FAILURES ===================================
_____________________________________ test _____________________________________

capsys = <_pytest.capture.CaptureFixture object at 0x7f15ca2cd320>

    def test(capsys):
>       sys.stdout.buffer.write(b'hi\n')
E       AttributeError: '_io.StringIO' object has no attribute 'buffer'

test.py:5: AttributeError
=========================== 1 failed in 0.01 seconds ===========================
```

### PR DESCRIPTION
**Pull request by asottile**
Replay of  #2265 against origin/features

### CODE REVIEW

**Comment by asottile** [comment]
See discussion in #2265

File: testing/test_capture.py [hunk 1]
```diff
@@ -642,6 +642,13 @@ def test_unicode_and_str_mixture(self):
             f.close()
             assert isinstance(s, unicode)
 
+    def test_write_bytes_to_buffer(self):
+        f = capture.CaptureIO()
+        # In python3, stdout / stderr are text io wrappers (exposing a buffer
+        # property of the underlying bytestream).
```
Comment (nicoddemus) [inline testing/test_capture.py]: I would rather at this point to be explicit about the expected behavior in Python 3:

```python
@pytest.mark.skipif(sys.version_info[0] == 2, 'python 3 only behavior')
def test_write_bytes_to_buffer(self):
    """In python3, stdout / stderr are text io wrappers (exposing a buffer
    property of the underlying bytestream). See issue #1407
    """
    f = capture.CaptureIO()
    f.buffer.write(b'foo\r\n')
    assert f.getvalue() == 'foo\n'
```

(I also added a `\r\n` there to ensure we are not messing with how newlines are treated)
    
```

**Comment by nicoddemus** [comment]
Thanks for following up @asottile!

Looking over at #1407, I noticed this [comment](https://github.com/pytest-dev/pytest/issues/1407#issuecomment-260144848) from @RonnyPfannschmidt. @RonnyPfannschmidt you still think the same way?

File: testing/test_capture.py (hunk 1 above)
Comment (asottile) [inline testing/test_capture.py]: Done!

**Comment by asottile** [comment]
Looks like travis-ci is only failing due to the nightly

**Comment by nicoddemus** [comment]
Yes, and the same failures are happening on `master` for the `nightly` job. The AppVeyor failure seemed to be a fluke of the CI, I have triggered it again.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus i still think the capture objects should be simple/simplified, but since we have no way to properly do any cases, this pr adds value in any case

**Comment by nicoddemus** [comment]
Thanks @RonnyPfannschmidt, merging it then.

Thanks @asottile again for the contribution!

**Pull request merged**